Source/r_main.c

```diff
@@ -305,19 +314,83 @@ static void R_InitTextureMapping (void)
 
 #define DISTMAP 2
 
+boolean smoothlight;
+
 void R_InitLightTables (void)
 {
-  int i;
+  int i, cm;
     
+  if (c_scalelight)
+  {
+    for (cm = 0; cm < numcolormaps; ++cm)
+    {
+      for (i = 0; i < LIGHTLEVELS; ++i)
+        free(c_scalelight[cm][i]);
+
+      free(c_scalelight[cm]);
+    }
+    free(c_scalelight);
+  }
+
+  if (scalelightfixed)
+  {
+    free(scalelightfixed);
+  }
+
+  if (c_zlight)
+  {
+    for (cm = 0; cm < numcolormaps; ++cm)
+    {
+      for (i = 0; i < LIGHTLEVELS; ++i)
+        free(c_zlight[cm][i]);
+
+      free(c_zlight[cm]);
+    }
+    free(c_zlight);
+  }
+
+  if (smoothlight)
+  {
+      LIGHTLEVELS = 32;
+      LIGHTSEGSHIFT = 3;
+      LIGHTBRIGHT = 2;
+      MAXLIGHTSCALE = 48;
+      LIGHTSCALESHIFT = 12;
+      MAXLIGHTZ = 1024;
+      LIGHTZSHIFT = 17;
+  }
+  else
+  {
+      LIGHTLEVELS = 16;
+      LIGHTSEGSHIFT = 4;
+      LIGHTBRIGHT = 1;
+      MAXLIGHTSCALE = 48;
+      LIGHTSCALESHIFT = 12;
+      MAXLIGHTZ = 128;
+      LIGHTZSHIFT = 20;
+  }
+
+  scalelightfixed = malloc(MAXLIGHTSCALE * sizeof(*scalelightfixed));
```

Comment:
Better use stdlib `(malloc)()`/`(free)()` in this code? Not sure...

Reply:
Lee Killough used zone memory for c_zlight c_scalelight (colormaps). I think we should remove memory macros, it's not that hard.

Reply:
Hm, we could define the `malloc` macro to something non-existing and then fix all the places in the code where it fails to build to explicitly call `Z_Malloc`, for example. Indeed, I am all for a clean separation between `Z_Malloc` and stdlib `malloc` in the code, but was too afraid to even start the transition. 😉

Reply:
When I tried to do this in dsda-doom I ran into some issues and ultimately put it off. There's no automatic way to know if file A uses Z_malloc and file B uses std free on the same data (which would crash) at compile time as far as I know. You'd need to check all the free calls I guess, see which ones aren't isolated to a file, and confirm consistency.

Reply:
Or would making the macro null actually detect this? 🤔

Reply:
So, I just did this and was confronted with a shitload of warnings pointing to where exactly functions were called that were aliased to their Z-pendants before.

```patch
--- a/Source/z_zone.h
+++ b/Source/z_zone.h
@@ -88,11 +88,11 @@ void Z_DumpHistory(char *);
 #undef calloc
 #undef strdup

-#define malloc(n)          Z_Malloc(n,PU_STATIC,0)
-#define free(p)            Z_Free(p)
-#define realloc(p,n)       Z_Realloc(p,n,PU_STATIC,0)
-#define calloc(n1,n2)      Z_Calloc(n1,n2,PU_STATIC,0)
-#define strdup(s)          Z_Strdup(s,PU_STATIC,0)
+#define malloc(n)          X_Malloc(n,PU_STATIC,0)
+#define free(p)            X_Free(p)
+#define realloc(p,n)       X_Realloc(p,n,PU_STATIC,0)
+#define calloc(n1,n2)      X_Calloc(n1,n2,PU_STATIC,0)
+#define strdup(s)          X_Strdup(s,PU_STATIC,0)

 // dprintf() is already declared in <stdio.h>, define it out of the way
 #define dprintf doomprintf
```

Reply:
There is an gcc -E option, it only runs the preprocessor. Thus, we can comment out all `#include` except z_zone.h. I tried, but it might be easier to do it manually.

Reply:
Example:
```c
[1/51] Building C object Source/CMakeFiles/woof.dir/i_glob.c.obj
FAILED: Source/CMakeFiles/woof.dir/i_glob.c.obj 
D:\msys64\mingw32\bin\cc.exe -DHAVE_FLUIDSYNTH -DMINIZ_NO_ARCHIVE_APIS -DMINIZ_NO_STDIO -DMINIZ_NO_TIME -DRANGECHECK -Ibuild/Source/.. -Itextscreen/. -Iopl/. -isystem D:/msys64/mingw32/include/SDL2 -O2 -g -DNDEBUG -Wdeclaration-after-statement -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Werror=int-conversion -Wformat=2 -Wnull-dereference -Wredundant-decls -Wrestrict -Wall -Werror -Dmain=SDL_main -MD -MT Source/CMakeFiles/woof.dir/i_glob.c.obj -MF Source\CMakeFiles\woof.dir\i_glob.c.obj.d -o Source/CMakeFiles/woof.dir/i_glob.c.obj -c Source/i_glob.c
In file included from Source/doomdef.h:34,
                 from Source/m_misc.h:34,
                 from Source/i_glob.c:24:
Source/i_glob.c: In function 'FreeStringList':
Source/z_zone.h:92:28: error: implicit declaration of function 'X_Free'; did you mean 'Z_Free'? [-Werror=implicit-function-declaration]
   92 | #define free(p)            X_Free(p)
      |                            ^~~~~~
Source/i_glob.c:103:5: note: in expansion of macro 'free'
  103 |     free(globs);
      |     ^~~~
Source/i_glob.c: In function 'I_StartMultiGlob':
Source/z_zone.h:91:28: error: implicit declaration of function 'X_Malloc'; did you mean 'Z_Malloc'? [-Werror=implicit-function-declaration]
   91 | #define malloc(n)          X_Malloc(n,PU_STATIC,0)
      |                            ^~~~~~~~
Source/i_glob.c:114:13: note: in expansion of macro 'malloc'
  114 |     globs = malloc(sizeof(char *));
      |             ^~~~~~
Source/i_glob.c:114:11: error: assignment to 'char **' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  114 |     globs = malloc(sizeof(char *));
      |           ^
In file included from Source/doomdef.h:34,
                 from Source/m_misc.h:34,
                 from Source/i_glob.c:24:
Source/z_zone.h:93:28: error: implicit declaration of function 'X_Realloc'; did you mean 'Z_Realloc'? [-Werror=implicit-function-declaration]
   93 | #define realloc(p,n)       X_Realloc(p,n,PU_STATIC,0)
      |                            ^~~~~~~~~
Source/i_glob.c:133:21: note: in expansion of macro 'realloc'
  133 |         new_globs = realloc(globs, sizeof(char *) * (num_globs + 1));
      |                     ^~~~~~~
Source/i_glob.c:133:19: error: assignment to 'char **' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  133 |         new_globs = realloc(globs, sizeof(char *) * (num_globs + 1));
      |                   ^
Source/i_glob.c:144:12: error: assignment to 'glob_t *' {aka 'struct glob_s *'} from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  144 |     result = malloc(sizeof(glob_t));
      |            ^
Source/i_glob.c: In function 'ReadAllFilenames':
Source/i_glob.c:283:25: error: assignment to 'char **' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  283 |         glob->filenames = realloc(glob->filenames,
      |                         ^
cc1.exe: all warnings being treated as errors
```